R/layers2traces.R

```diff
@@ -624,9 +624,120 @@ to_basic.GeomQuantile <- function(data, prestats_data, layout, params, p, ...){
+
+  inter <- intersect(names(data), paste0("subgroup", c("", 2:3)))
+  if(length(inter)) pars[inter] <- inter
+  
+  do.call(treemapify:::treemapify, pars)
+}
+
+#' @export 
+to_basic.GeomTreemapText <- function(data, prestats_data, layout, params, p, ...){
+  data <- tree_transform(data, params)
+
+  if(any(grepl("subgroup", params))) 
+
+  data$size <- with(data, 2*(xmax - xmin)/strwidth(label, units = "figure"))
+  data[, c("x", "y", "textposition")] <- with(data, list(x = (xmin+xmax)/2, y=(ymin+ymax)/2 , textposition = params$place))
+  #data[, c("x", "y", "hjust", "vjust")] <- with(data, place_to_coords(xmin, xmax, ymin, ymax, params$place))
+  #data[, c("x", "y")] <- with(data, list(x = (xmax+xmin)/2, y = if(any(grepl("subgroup", params))) ymax - strheight(label, units="figure")*.5*size else (ymax+ymin)/2 ) )
+  data$colour <- params$colour
+  data$fontface <- params$fontface
+  
+  prefix_class(data, "GeomText")
+}
+#place_to_coords <- function(xmin, xmax, ymin, ymax, place){
+#  #width <- strwidth(label)
+#  #height <- strheight(label)
+#  switch(place, 
+#    "bottom" = list(y = (ymax+ymin)/2, x = (xmin+xmax)/2, hjust=0, vjust=0),
+#    "right" = list(y = xmax, y = (ymin+ymax)/2, hjust=0, vjust=.5),
+#    "middle" = list(y = (xmax+xmax)/2, y = (ymin+ymax)/2, hjust=.5, vjust=.5),
+#    "left" = list(y = xmin, y = (ymin+ymax)/2, hjust = .5, vjust=.5),
+#    "top" =  list(y = ymax, x = (xmin+xmax)/2, vjust=0, hjust=.5),
+#  )
+#}
+treesubgroup_transform <- function(data, params){
+
+  pars <- params[c("fixed", "layout", "start")]
+  pars$area <- "area"
+
+  levels <- paste0("subgroup", c("", 2:3))
+  
+  levels <- levels[1:which(levels == params$level)]
+
+
+  bys <- lapply(levels, function(x) data[[x]])
+  areasums <- aggregate(data$area, by = bys, FUN = sum)
+  names(areasums) <- c(levels, "area")
+  for (aesthetic in setdiff(names(data), names(areasums))) {
+    values <- data[[aesthetic]]
+    names(values) <- data[[params$level]]
+    areasums[aesthetic] <- values[as.character(areasums[[params$level]])]
+  }
+
+  
+  pars$data <- areasums
+  if(length(levels) > 1) pars[head(levels, -1)] <- head(levels, -1)
+
+  do.call(treemapify:::treemapify, pars)
+
+}
+
+#' @export 
+to_basic.GeomSubgroupBorder <- function(data, prestats_data, layout, params, p, ...){
+  prefix_class(to_basic.GeomRect(treesubgroup_transform(data, params)), "GeomPath")
+}
+#' @export 
+to_basic.GeomSubgroupText <- function(data, prestats_data, layout, params, p, ...){
+  data <- treesubgroup_transform(data, params)
+  names(data)[names(data) == params$level] <- "label"
+
+  data$size <- with(data, 3*(xmax - xmin)/strwidth(label, units = "figure"))
+  #data[, c("x", "y")] <- with(data, list( x = (xmin+xmax)/2, y = (ymin+ymax)/2 ))
+  data[, c("x", "y", "textposition")] <- with(data, list(x = (xmin+xmax)/2, y=(ymin+ymax)/2 , textposition = params$place))
+
+  data$colour <- params$colour
+  data$fontface <- params$fontface
+  prefix_class(data, "GeomText")
+}
```

Comment:
Is this code being lifted from somewhere else? If yes, please add proper attribution. Also, if `:::` is absolutely necessary, you'll have to use `getFromNamespace()` instead.

Reply:
the code was adapted from the `treemapify` package

Reply:
This could be problematic considering that `treemapify` is GPL licensed and plotly is MIT licensed https://github.com/wilkox/treemapify/blob/master/DESCRIPTION#L23

Reply:
tbh only some parts of the logic is transfered. I didn't actually copy/paste the code.